include/boost/math/special_functions/jacobi_theta.hpp

```diff
@@ -0,0 +1,665 @@
+
+#ifndef BOOST_MATH_JACOBI_THETA_HPP
+#define BOOST_MATH_JACOBI_THETA_HPP
+
+#include <boost/math/tools/complex.hpp>
+#include <boost/math/tools/precision.hpp>
+#include <boost/math/tools/promotion.hpp>
+#include <boost/math/policies/error_handling.hpp>
+#include <boost/math/constants/constants.hpp>
+
+namespace boost{ namespace math{
+
+// Simple functions - parameterized by q
+template <class T>
+inline T jacobi_theta1(T z, T q);
+template <class T>
+inline T jacobi_theta2(T z, T q);
+template <class T>
+inline T jacobi_theta3(T z, T q);
+template <class T>
+inline T jacobi_theta4(T z, T q);
+
+// Simple functions - parameterized by tau (assumed imaginary)
+// q = exp(iπτ)
+// tau = -log(q)/π
+template <class T>
+inline T jacobi_theta1tau(T z, T tau);
+template <class T>
+inline T jacobi_theta2tau(T z, T tau);
+template <class T>
+inline T jacobi_theta3tau(T z, T tau);
+template <class T>
+inline T jacobi_theta4tau(T z, T tau);
+
+// Minus one versions for small q / large tau
+template <class T>
+inline T jacobi_theta3m1(T z, T q);
+template <class T>
+inline T jacobi_theta4m1(T z, T q);
+template <class T>
+inline T jacobi_theta3m1tau(T z, T tau);
+template <class T>
+inline T jacobi_theta4m1tau(T z, T tau);
+
+// Policied versions - parameterized by q
+template <class RealType, class Policy>
+inline RealType jacobi_theta1(RealType z, RealType q, const Policy& pol);
+template <class RealType, class Policy>
+inline RealType jacobi_theta2(RealType z, RealType q, const Policy& pol);
+template <class RealType, class Policy>
+inline RealType jacobi_theta3(RealType z, RealType q, const Policy& pol);
+template <class RealType, class Policy>
+inline RealType jacobi_theta4(RealType z, RealType q, const Policy& pol);
+
+// Policied versions - parameterized by tau
+template <class RealType, class Policy>
+inline RealType jacobi_theta1tau(RealType z, RealType tau, const Policy& pol);
+template <class RealType, class Policy>
+inline RealType jacobi_theta2tau(RealType z, RealType tau, const Policy& pol);
+template <class RealType, class Policy>
+inline RealType jacobi_theta3tau(RealType z, RealType tau, const Policy& pol);
+template <class RealType, class Policy>
+inline RealType jacobi_theta4tau(RealType z, RealType tau, const Policy& pol);
+
+// Policied m1 functions
+template <class RealType, class Policy>
+inline RealType jacobi_theta3m1(RealType z, RealType q, const Policy& pol);
+template <class RealType, class Policy>
+inline RealType jacobi_theta4m1(RealType z, RealType q, const Policy& pol);
+template <class RealType, class Policy>
+inline RealType jacobi_theta3m1tau(RealType z, RealType tau, const Policy& pol);
+template <class RealType, class Policy>
+inline RealType jacobi_theta4m1tau(RealType z, RealType tau, const Policy& pol);
+
+// Internal convergence criterion:
+// If the delta is more than an absolute epsilon, keep going.
+// If the delta is more than an relative epsilon, and the denominator
+// is non-zero, keep going.
```

Comment:
I'm not sure this comment is very informative; I was hoping for something like "expected error is cond(f; x)eps/2, so when delta <..." and then some perturbative analysis.

Reply:
The comment represents the extent of my thinking and, well, education. I'm happy to incorporate a more formal convergence analysis but it's simply outside my ken.

Reply:
lol, alright np. I'll see if I can get a minute to work out some error analysis for these functions.

Reply:
IMO there's no need to require `delta < eps`, normally we would just use `abs(result * eps) > abs(delta)` as if that's true, then adding `delta` can have no effect on `result`.  The only exception would be if `delta` can grow as well as shrink.

Reply:
The sign of `delta` can alternate, which could theoretically put `result` at 0 after two iterations. So I wanted to handle that somehow, and distinguish the situation from an underflow.

Reply:
The other point that you are doing a slow division instead of a faster multiplication still stands though. There is still a bit of thinking to do on this termination condition . . .

Reply:
Ok, I think I've figured this out. The reason that the sign of `delta` can alternate is because you're including the oscillatory part in it (sin((2n+1)x) or cos((2n+1)x)). I think you should only be testing the magnitude of q^n^2, because the magnitude of complex phase of the original complex valued Jacobi function is always unity. As an aside, the auxiliary functions look much more difficult to implement individually than just using the original Jacobi function in the complex plane. . .

Reply:
The convergence tests operate on `q_n` and do not involve the oscillatory part. My original intention was to operate in the complex plane, but the API got too complicated when trying to support arbitrary numeric types.

Reply:
Wait then aren't all these calls to `abs` superfluous?

Reply:
Sorry, I should have said: the numerator is `q_n` (always positive) but the denominator is `result` (which may oscillate). It would probably be better to change the denominator to `previous_q_n`.

Reply:
I personally might track \sum q^{n^2} since it provides a nice L1 bound on the sum.

Reply:
I'm not sure the sum will work because of the alternating signs. In addition to the oscillating component there is the (-1)^n in two of the formulas. With q close to 1, the straight sum could be large relative to the final result, and so an error that is small relative to the sum may be large relative to the result.

Reply:
I have updated the convergence criterion to compare `q_n` to the previous one. Everything including ULP plots look pretty much the same. Let me know what you think.